contracts/MechMarketplace.sol

```diff
@@ -199,103 +248,46 @@ contract MechMarketplace {
-        
-        emit MinMaxResponseTimeoutUpdated(newMinResponseTimeout, newMaxResponseTimeout);
-    }
-
-    /// @dev Sets mech registration status.
-    /// @param mech Mech address.
-    /// @param status True, if registered, false otherwise.
-    function setMechRegistrationStatus(address mech, bool status) external {
-        // Check for the agent mech factory access, contract ownership or mech itself (when changing its marketplace)
-        if (msg.sender != factory && msg.sender != owner && msg.sender != mech) {
-            revert OwnerOnly(msg.sender, owner);
+    /// @dev Checks agent mech for contract validity.
+    /// @param mech Agent mech address.
+    /// @param mechServiceId Mech operator service Id.
+    function checkMech(address mech, uint256 mechServiceId) public view {
+        // Check that the mech address corresponds to the authorized bytecode hash
+        bytes32 mechHash = keccak256(mech.code);
+        if (mechHash != mechBytecodeHash) {
+            revert UnauthorizedAccount(mech);
         }
 
-        // Prevent mech calling this function by exec() and trying to whitelist itself
-        if (msg.sender == mech && status) {
-            revert UnauthorizedAccount(msg.sender);
+        // Check if the mech service is staked
+        IStaking.StakingState state = IStaking(mechStakingInstance).getStakingState(mechServiceId);
+        if (state != IStaking.StakingState.Staked) {
+            revert ServiceNotStaked(mechStakingInstance, mechServiceId);
         }
 
-        // Check that mech is a contract
-        if (mech.code.length == 0) {
-            revert NotContract(mech);
+        // Get the staked service info
+        IStaking.ServiceInfo memory serviceInfo = IStaking(mechStakingInstance).getServiceInfo(mechServiceId);
+        // Check that staked service multisig is the priority mech operator
+        if (!IMech(mech).isOperator(serviceInfo.multisig)) {
+            revert UnauthorizedAccount(mech);
         }
-
-        mapMechRegistrations[mech] = status;
-        emit MechRegistrationStatusChanged(mech, status);
     }
 
     /// @dev Registers a request.
     /// @notice The request is going to be registered by a specified priority agent mech.
     /// @param data Self-descriptive opaque data-blob.
     /// @param priorityMech Address of a priority mech.
+    /// @param priorityMechServiceId Priority mech operator service Id.
     /// @param responseTimeout Relative response time in sec.
+    /// @param requesterStakingInstance Staking instance of a service whose multisig posts a request.
+    /// @param requesterServiceId Corresponding service Id in the staking contract.
     /// @return requestId Request Id.
     function request(
         bytes memory data,
         address priorityMech,
-        uint256 responseTimeout
+        uint256 priorityMechServiceId,
+        uint256 responseTimeout,
+        address requesterStakingInstance,
+        uint256 requesterServiceId
```

Comment:
In order to check the priority mech validity and requester services validity (aka traders), their service Ids need to be supplied. Also, for the requester the corresponding service staking instance is needed.